src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,714 @@
+        internal
+        virtual
+        returns (uint256 amountIn)
+    {
+        uint256 price;
+        if (inToken == token0) {
+            require(outToken == token1, "ARM: Invalid out token");
+            price = traderate0;
+        } else if (inToken == token1) {
+            require(outToken == token0, "ARM: Invalid out token");
+            price = traderate1;
+        } else {
+            revert("ARM: Invalid in token");
+        }
+        amountIn = ((amountOut * PRICE_SCALE) / price) + 1; // +1 to always round in our favor
+
+        // Transfer the input tokens from the caller to this ARM contract
+        _transferAssetFrom(address(inToken), msg.sender, address(this), amountIn);
+
+        // Transfer the output tokens to the recipient
+        _transferAsset(address(outToken), to, amountOut);
+    }
+
+    /**
+     * @notice Set exchange rates from an operator account from the ARM's perspective.
+     * If token 0 is WETH and token 1 is stETH, then both prices will be set using the stETH/WETH price.
+     * @param buyT1 The price the ARM buys Token 1 (stETH) from the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the sell price.
+     * @param sellT1 The price the ARM sells Token 1 (stETH) to the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the buy price.
+     */
+    function setPrices(uint256 buyT1, uint256 sellT1) external onlyOperatorOrOwner {
+        // Ensure buy price is always below past sell prices
+        require(sellT1 >= crossPrice, "ARM: sell price too low");
+        require(buyT1 < crossPrice, "ARM: buy price too high");
+
+        traderate0 = PRICE_SCALE * PRICE_SCALE / sellT1; // base (t0) -> token (t1);
+        traderate1 = buyT1; // token (t1) -> base (t0)
+
+        emit TraderateChanged(traderate0, traderate1);
+    }
+
+    /**
+     * @notice set the price that buy and sell prices can not cross.
+     * That is, the buy prices must be below the cross price
+     * and the sell prices must be above the cross price.
+     * If the cross price is being lowered, there can not be any base assets in the ARM. eg stETH.
+     * The base assets should be sent to the withdrawal queue before the cross price can be lowered.
+     * The cross price can be increased with assets in the ARM.
+     * @param newCrossPrice The new cross price scaled to 36 decimals.
+     */
+    function setCrossPrice(uint256 newCrossPrice) external onlyOwner {
+        require(newCrossPrice >= PRICE_SCALE - MAX_CROSS_PRICE_DEVIATION, "ARM: cross price too low");
+        require(newCrossPrice <= PRICE_SCALE, "ARM: cross price too high");
+
+        // If the new cross price is lower than the current cross price
+        if (newCrossPrice < crossPrice) {
+            // Check there is not a significant amount of base assets in the ARM
```

Comment:
⚪️ This comment and the one above comment tells what we are doing, but not why we are doing it, which is the hard one to guess from the code.

Reply:
I've updated the comment to make it clearer